torchaudio/prototype/transforms/_transforms.py

```diff
@@ -384,3 +385,100 @@ def forward(self, waveform: torch.Tensor) -> torch.Tensor:
+    """
+
+    def __init__(self, orig_freq, factor) -> None:
+        super().__init__()
+
+        self.orig_freq = orig_freq
+        self.factor = factor
+
+        self.source_sample_rate, self.target_sample_rate = _source_target_sample_rate(orig_freq, factor)
+        self.resampler = Resample(orig_freq=self.source_sample_rate, new_freq=self.target_sample_rate)
+
+    def forward(self, waveform, lengths) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Args:
+            waveform (torch.Tensor): Input signals, with shape `(..., time)`.
+            lengths (torch.Tensor): Valid lengths of signals in ``waveform``, with shape `(...)`.
+
+        Returns:
+            (torch.Tensor, torch.Tensor):
+                torch.Tensor
+                    Speed-adjusted waveform, with shape `(..., new_time).`
+                torch.Tensor
+                    Valid lengths of signals in speed-adjusted waveform, with shape `(...)`.
+        """
+        return (
+            self.resampler(waveform),
+            torch.ceil(lengths * self.target_sample_rate / self.source_sample_rate).to(lengths.dtype),
+        )
+
+
+class SpeedPerturbation(torch.nn.Module):
+    r"""Applies the speed perturbation augmentation introduced in
+    *Audio augmentation for speech recognition* :cite:`ko15_interspeech`. For a given input,
+    with probability ``perturb_prob``, the module samples a speed-up factor from ``factors``
+    uniformly at random and adjusts the speed of the input by that factor.
+
+    .. devices:: CPU CUDA
+
+    .. properties:: Autograd TorchScript
+
+    Args:
+        orig_freq (int): Original frequency of the signals in ``waveform``.
+        factors (Sequence[float]): Factors by which to adjust speed of input. Values greater than 1.0
+            compress ``waveform`` in time, whereas values less than 1.0 stretch ``waveform`` in time.
+        perturb_prob (float): Probability with which to adjust speed of input.
+    """
+
+    def __init__(self, orig_freq: int, factors: Sequence[float], perturb_prob: float) -> None:
+        super().__init__()
+
+        self.perturb_prob = torch.tensor(perturb_prob)
+        self.speeders = torch.nn.ModuleList([Speed(orig_freq=orig_freq, factor=factor) for factor in factors])
+
+    def forward(self, waveform: torch.Tensor, lengths: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Args:
+            waveform (torch.Tensor): Input signals, with shape `(..., time)`.
+            lengths (torch.Tensor): Valid lengths of signals in ``waveform``, with shape `(...)`.
+
+        Returns:
+            (torch.Tensor, torch.Tensor):
+                torch.Tensor
+                    Speed-adjusted waveform, with shape `(..., new_time).`
+                torch.Tensor
+                    Valid lengths of signals in speed-adjusted waveform, with shape `(...)`.
+        """
+
+        if torch.bernoulli(self.perturb_prob):
```

Comment:
it looks like all waveforms in a batch will be either all perturbed together or all kept original? can we apply the bernoulli factor on each waveform instead? that'll add more randomness in the data especially when batch size is large.

Reply:
ok I agree. I don't it worth the performance degradation. let's keep it the current way. BTW I think it can also be achieved by inserting "1"s in the "factors" sequence, e.g. if we want perturb 40% data, then we just make sure 60% of the elements in "factors" are "1"s and use perturb_prob=1.

Reply:
that's a good idea. it wasn't the api that was originally proposed, but this seems cleaner and clearer